Lookup DOI (sometimes) freezes JabRef

New in 4.0 Development version!

With reverted log addition in FX thread:
Infinite loop?!
```
Opening: C:\Users\stfn\Desktop\Vorlesungen\Promotion\Paper\master.bib
Opening: C:\Users\stfn\Downloads\01 - TEMP\test.bib
Problem fetching DOI
org.jabref.logic.importer.FetcherException: An I/O exception occurred
	at org.jabref.logic.importer.IdParserFetcher.findIdentifier(IdParserFetcher.java:92)
	at org.jabref.gui.entryeditor.FieldExtraComponents.lambda$getDoiExtraComponent$3(FieldExtraComponents.java:186)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 500 for URL: http://api.crossref.org/works?query.title=A+break+in+the+clouds%3A+towards+a+cloud+definition&query.author=Vaquero%2C+Luis+M+and+Rodero-Merino%2C+Luis+and+Caceres%2C+Juan+and+Lindner%2C+Maik&filter=from-pub-date%3A2008&rows=20&offset=0
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.net.URL.openStream(Unknown Source)
	at org.jabref.logic.importer.IdParserFetcher.findIdentifier(IdParserFetcher.java:74)
	... 37 more

No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
```

Also happens for `get BibTeX data from DOI`:
```
Could not copy input
java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:152)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:169)
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:55)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:47)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:22)
	at javax.swing.SwingWorker$1.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at javax.swing.SwingWorker.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.net.HttpURLConnection.getResponseCode(Unknown Source)
	at org.jabref.logic.net.URLDownload.openConnection(URLDownload.java:297)
	... 11 more

Info cannot be found
org.jabref.logic.importer.FetcherException: Invalid URL
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:65)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:47)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:22)
	at javax.swing.SwingWorker$1.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at javax.swing.SwingWorker.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:152)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:169)
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:55)
	... 8 more
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.net.HttpURLConnection.getResponseCode(Unknown Source)
	at org.jabref.logic.net.URLDownload.openConnection(URLDownload.java:297)
	... 11 more

Could not copy input
java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:152)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:169)
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:55)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:47)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:22)
	at javax.swing.SwingWorker$1.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at javax.swing.SwingWorker.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.net.HttpURLConnection.getResponseCode(Unknown Source)
	at org.jabref.logic.net.URLDownload.openConnection(URLDownload.java:297)
	... 11 more

Info cannot be found
org.jabref.logic.importer.FetcherException: Invalid URL
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:65)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:47)
	at org.jabref.gui.mergeentries.FetchAndMergeWorker.doInBackground(FetchAndMergeWorker.java:22)
	at javax.swing.SwingWorker$1.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at javax.swing.SwingWorker.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:152)
	at org.jabref.logic.net.URLDownload.asString(URLDownload.java:169)
	at org.jabref.logic.importer.fetcher.DoiFetcher.performSearchById(DoiFetcher.java:55)
	... 8 more
Caused by: java.io.IOException: Server returned HTTP response code: 503 for URL: http://data.crossref.org/10.1145%2F1496091.1496100
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
	at java.net.HttpURLConnection.getResponseCode(Unknown Source)
	at org.jabref.logic.net.URLDownload.openConnection(URLDownload.java:297)
	... 11 more

No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
No localized message exception message defined. Falling back to getMessage().
````

--- a/src/main/java/org/jabref/JabRefException.java
+++ b/src/main/java/org/jabref/JabRefException.java
@@ -5,9 +5,8 @@ import org.apache.commons.logging.LogFactory;
 
 public class JabRefException extends Exception {
 
-    private String localizedMessage;
-
     private static final Log LOGGER = LogFactory.getLog(JabRefException.class);
+    private String localizedMessage;
 
     public JabRefException(String message) {
         super(message);
@@ -34,7 +33,7 @@ public class JabRefException extends Exception {
     @Override
     public String getLocalizedMessage() {
         if (localizedMessage == null) {
-            LOGGER.warn("No localized message exception message defined. Falling back to getMessage().");
+            LOGGER.debug("No localized message exception message defined. Falling back to getMessage().");
             return getMessage();
         } else {
             return localizedMessage;

--- a/src/main/java/org/jabref/gui/logging/GuiAppender.java
+++ b/src/main/java/org/jabref/gui/logging/GuiAppender.java
@@ -2,6 +2,7 @@ package org.jabref.gui.logging;
 
 import java.io.Serializable;
 
+import org.jabref.gui.util.DefaultTaskExecutor;
 import org.jabref.logic.logging.LogMessages;
 
 import org.apache.logging.log4j.core.Filter;
@@ -43,6 +44,6 @@ public class GuiAppender extends AbstractAppender {
      */
     @Override
     public void append(LogEvent event) {
-        LogMessages.getInstance().add(event);
+        DefaultTaskExecutor.runInJavaFXThread(() -> LogMessages.getInstance().add(event));
     }
 }

--- a/src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java
+++ b/src/main/java/org/jabref/gui/util/DefaultTaskExecutor.java
@@ -19,6 +19,21 @@ public class DefaultTaskExecutor implements TaskExecutor {
 
     private static final Log LOGGER = LogFactory.getLog(DefaultTaskExecutor.class);
 
+    public static <V> V runInJavaFXThread(Callable<V> callable) {
+        FutureTask<V> task = new FutureTask<>(callable);
+        Platform.runLater(task);
+        try {
+            return task.get();
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.error(e);
+            return null;
+        }
+    }
+
+    public static void runInJavaFXThread(Runnable runnable) {
+        Platform.runLater(runnable);
+    }
+
     @Override
     public <V> void execute(BackgroundTask<V> task) {
         new Thread(getJavaFXTask(task)).start();
@@ -54,16 +69,4 @@ public class DefaultTaskExecutor implements TaskExecutor {
             return new Exception(throwable);
         }
     }
-
-    public static <V> V runInJavaFXThread(Callable<V> callable) {
-        FutureTask<V> task = new FutureTask<>(callable);
-        Platform.runLater(task);
-        try {
-            return task.get();
-        } catch (InterruptedException | ExecutionException e) {
-            LOGGER.error(e);
-            return null;
-        }
-    }
-
 }

